### PR TITLE
Waterfall Fix for Initial Load

### PR DESCRIPTION
**Pull request by WhiteNervosa**
This fixes the [waterfall](https://i.imgur.com/8LVbbps.png) when chat initially loads. For UX, chat will load twice as fast and the further you are away from the server the more noticeable it is because TTFB now only matters for the last asset loaded.

```if(viewerStates) ``` is then necessary for Dev Chat.

@babel/polyfill is needed to allow async/await when it initializes.

On the right is the waterfall, currently, with assets and API requests being done and waiting for the prior to finish before starting the next. On the left is them all being requested at the same time. Once the websocket loads the user can chat. Since most of the data is quite small, most of the time waiting is the TTFB, by doing all the requests at once the TTFB is only as slow as the last request, effectively.

dg2xovi 1

TLDR ( or because I cannot explain things coherently ) - Loading everything at the same time rather than one by one in a queue.

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,3 +1,10 @@
 {
```
Comment (WhiteNervosa) [inline .babelrc:1]: This is necessary for the async/await initialization

File: .babelrc [hunk 2]
```diff
@@ -1,3 +1,10 @@
...
-    "presets": ["@babel/preset-env"]
+    "presets": [
```
Comment (WhiteNervosa) [inline .babelrc:2]: This is necessary for the async/await initialization

File: assets/chat.js [hunk 3]
```diff
@@ -1,30 +1,18 @@
...
-            .withHistory(history)
-            .withWhispers()
-            .connect(WEBSOCKET_URI))
-);
+const init = async v => {
+    const userInfo = async v => await new Promise(res => $.getJSON(`${API_URI}/api/chat/me`).done(res).fail(() => res(null)));
+    const history = async v => await new Promise(res => $.getJSON(`${API_URI}/api/chat/history`).done(res).fail(() => res(null)));
+    const viewerStates = async v => await new Promise(res => $.getJSON(`${API_URI}/api/chat/viewer-states`).done(res).fail(() => res(null)));
+
+    const promiseSync = await Promise.all([userInfo(), history(), viewerStates()]);
```
Comment (WhiteNervosa) [inline assets/chat.js:9]: Instead of loading user info, then history, etc. they are loaded all at once. Then, in the second batch user settings, whispers, are all loaded. The two need to be separate as the second set depends on the information from the first.

The dev build worked without setting ```window.__chat___``` but just in case it is set for mobile it was set again, here.

File: package.json [hunk 4]
```diff
@@ -37,6 +37,7 @@
   },
   "devDependencies": {
     "@babel/core": "^7.4.0",
+    "@babel/polyfill": "^7.10.4",
```
Comment (WhiteNervosa) [inline package.json:40]: Polyfill is necessary for the async/await initialization

File: assets/chat/js/chat.js [hunk 5]
```diff
@@ -658,13 +658,13 @@ class Chat {
         return this;
     }
 
-    withViewerStates(viewerStates) {
-        viewerStates.forEach(state => this.onVIEWERSTATE(state));
+    async withViewerStates(viewerStates) {
+        if(viewerStates) viewerStates.forEach(state => this.onVIEWERSTATE(state));
```
Comment (WhiteNervosa) [inline assets/chat/js/chat.js:662]: Dev Chat would error out with this async/await code, this check is to fix that. As history loads in Dev Chat, viewer states should load when given data and not null.

With Whispers doesn't have the data about whether the user is authenticated or not because it is all loaded at the same time, but there is no delay added for users who aren't signed in because a 503 error is returned and for users signed in it loads as normal.

Connect is changed to async to allow for .then

File: assets/chat.js [hunk 6]
```diff
@@ -1,30 +1,18 @@
...
-            .withGui()
-            .withViewerStates(viewerStates)
-            .withHistory(history)
-            .withWhispers()
-            .connect(WEBSOCKET_URI))
-);
+const init = async v => {
+    const userInfo = async v => await new Promise(res => $.getJSON(`${API_URI}/api/chat/me`).done(res).fail(() => res(null)));
+    const history = async v => await new Promise(res => $.getJSON(`${API_URI}/api/chat/history`).done(res).fail(() => res(null)));
+    const viewerStates = async v => await new Promise(res => $.getJSON(`${API_URI}/api/chat/viewer-states`).done(res).fail(() => res(null)));
```
Comment (slugalisk) [inline assets/chat.js:7]: the proposed changes just wrap the promises in more promises

the modern js would be something like 
```js
const [me, history, viewerStates] = await Promise.all([
  fetch(`${API_URI}/api/chat/me`).then(res => res.json()),
  fetch(`${API_URI}/api/chat/history`).then(res => res.json()),
  fetch(`${API_URI}/api/chat/viewer-states`).then(res => res.json()),
])
```
but neither of these improve the way the requests run... `$.getJSON` is already async and `$.when` is equivalent to `Promise.all`.

File: assets/chat.js [hunk 7]
```diff
@@ -1,30 +1,18 @@
...
+
+    const promiseSync = await Promise.all([userInfo(), history(), viewerStates()]);
+    
+    const chat = new Chat();
+    window.__chat__ = chat; // I do not know why this is needed but included it if an app depends on it
+    await chat.withGui()
+    .then(async() => await Promise.all([chat.withUserAndSettings(promiseSync[0]), chat.withEmotes(emotes), chat.withViewerStates(promiseSync[2]), chat.withHistory(promiseSync[1]), chat.withWhispers()]))
```
Comment (slugalisk) [inline assets/chat.js:14]: adding async to functions that do not perform async operations does not make them run in parallel.

Comment (WhiteNervosa) [reply assets/chat.js:14]: Pardon, you are correct and I knew about that but only an updated version of whispers would benefit and the rest would remain unaffected. I did not know about when doing the same.
